pkg/sql/show_fingerprints.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2017 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package sql
+
+import (
+	"fmt"
+	"strings"
+
+	"golang.org/x/net/context"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/parser"
+	"github.com/cockroachdb/cockroach/pkg/sql/privilege"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
+	"github.com/pkg/errors"
+)
+
+// ShowFingerprints statement fingerprints the data in each index of a table.
+// For each index, a full index scan is run to hash every row with the fnv64
+// hash. For the primary index, all table columns are included in the hash,
+// whereas for secondary indexes, the index cols + the primary index cols + the
+// STORING cols are included. The hashed rows are all combined with XOR using
+// distsql.
+//
+// Our hash functions expect input of type BYTES (or string but we use bytes
+// here), so we have to convert any datums that are not BYTES. This is currently
+// done by round tripping through the string representation of the column
+// (`::string::bytes`) and is an obvious area for improvement in the next
+// version.
+//
+// AS OF SYSTEM TIME is optional, but when set, uses the table definition and
+// data as of that the specified time.
+func (p *planner) ShowFingerprints(
+	ctx context.Context, n *parser.ShowFingerprints,
+) (planNode, error) {
+	ts := p.session.execCfg.Clock.Now()
+	if n.AsOf.Expr != nil {
+		evalCtx := p.session.evalCtx()
+		var err error
+		ts, err = EvalAsOfTimestamp(&evalCtx, n.AsOf, ts)
+		if err != nil {
+			return nil, err
+		}
+	}
+	p.txn.SetFixedTimestamp(ts)
```

Comment:
Ok I didn't realize you needed to set the timestamps.
In that case you should absolutely use the internal executor and a separate session/txn object. Otherwise this is going to trash the executor logic if someone ever runs this statement in a larger SQL transaction.